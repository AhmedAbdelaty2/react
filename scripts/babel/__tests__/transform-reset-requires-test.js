'use strict';

const babel = require('@babel/core');
const myPlugin = require('../transform-reset-requires'); // Adjust the path to where your plugin is located

const transform = code => {
  return babel.transform(code, {
    configFile: false,
    plugins: [
      // require.resolve('@babel/plugin-transform-modules-commonjs'),
      [
        myPlugin,
        {
          moduleNames: [
            'react',
            'react-dom',
            'react-dom/client',
            'internal-test-utils',
            'scheduler',
            'react-noop-renderer',
            'scheduler/unstable_mock',
          ],
        },
      ],
    ],
  }).code;
};

function normalizeIndent(strings) {
  const codeLines = strings[0].split('\n');
  const leftPadding = codeLines[1].match(/\s+/)[0];
  return codeLines.map(line => line.slice(leftPadding.length)).join('\n');
}

function formatSnapshot(inputCode) {
  return `### Before\n${inputCode}\n\n### After\n${transform(inputCode)}`;
}
expect.extend({
  toEqualIgnoringWhitespace(received, expected) {
    const normalizeWhitespace = str => str.replace(/\s+/g, ' ').trim();
    const normalizedReceived = normalizeWhitespace(received);
    const normalizedExpected = normalizeWhitespace(expected);

    if (normalizedReceived === normalizedExpected) {
      return {
        message: () =>
          `expected ${received} not to equal (ignoring whitespace) ${expected}`,
        pass: true,
      };
    } else {
      return {
        message: () =>
          `expected ${received} to equal (ignoring whitespace) ${expected}`,
        pass: false,
      };
    }
  },
});

describe('transformPlugin', () => {
  it("transforms require('react') correctly", () => {
    const inputCode = `const React = require('react');`;

    expect(formatSnapshot(inputCode)).toMatchSnapshot();
  });

  it("transforms require('react-dom') correctly", () => {
    const inputCode = `const ReactDOM = require('react-dom');`;

    expect(formatSnapshot(inputCode)).toMatchSnapshot();
  });

  it("transforms require('react-dom/client') correctly", () => {
    const inputCode = `const ReactDOMClient = require('react-dom/client');`;

    expect(formatSnapshot(inputCode)).toMatchSnapshot();
  });

  it('transforms act correctly', () => {
    const inputCode = `const act = require('internal-test-utils').act;`;

    expect(formatSnapshot(inputCode)).toMatchSnapshot();
  });

  it('transforms destructured imports correctly', () => {
    const inputCode = normalizeIndent`
      const React = require('react');
      const {startTransition, useDeferredValue} = React;
      const {
        waitFor,
        waitForAll,
        waitForPaint,
        waitForThrow,
        assertLog,
        act,
      } = require('internal-test-utils');
    `;

    expect(formatSnapshot(inputCode)).toMatchSnapshot();
  });

  it('transforms babel output correctly', () => {
    const inputCode = normalizeIndent`
    var React = require('react');var _React = React,
    startTransition = _React.startTransition,useDeferredValue = _React.useDeferredValue;
    var ReactNoop = require('react-noop-renderer');var _require = require('internal-test-utils'),waitFor = _require.waitFor,waitForAll = _require.waitForAll,waitForPaint = _require.waitForPaint,waitForThrow = _require.waitForThrow,assertLog = _require.assertLog;
    var act = require('internal-test-utils').act;
    `;

    expect(formatSnapshot(inputCode)).toMatchSnapshot();
  });

  it('transforms all imports correctly', () => {
    const inputCode = normalizeIndent`
      const React = require('react');
      const ReactDOM = require('react-dom');
      const ReactDOMClient = require('react-dom/client');
      const act = require('internal-test-utils').act;
    `;

    expect(formatSnapshot(inputCode)).toMatchSnapshot();
  });

  it('inserts before each at top', () => {
    const inputCode = normalizeIndent`
      const React = require('react');
      describe('test', () => {

      });
    `;

    expect(formatSnapshot(inputCode)).toMatchSnapshot();
  });

  it('does not transform non-configured modules', () => {
    const inputCode = `const moment = require('moment');`;
    const outputCode = transform(inputCode);

    expect(outputCode).toEqualIgnoringWhitespace(inputCode);
  });

  it('does not transform nested imports', () => {
    const inputCode = normalizeIndent`
      describe(() => {
        const React = require('react');
        const ReactDOM = require('react-dom');
        const ReactDOMClient = require('react-dom/client');
      });
    `;
    const outputCode = transform(inputCode);

    expect(outputCode).toEqualIgnoringWhitespace(inputCode);
  });

  it('does not transform lets in the existing pattern', () => {
    const inputCode = normalizeIndent`
    let React;
    let ReactDOM;
    let ReactDOMClient;
    beforeEach(() => {
      React = require('react');
      ReactDOM = require('react-dom');
      ReactDOMClient = require('react-dom/client');
    });`;
    const outputCode = transform(inputCode);

    expect(outputCode).toEqualIgnoringWhitespace(inputCode);
  });
});
