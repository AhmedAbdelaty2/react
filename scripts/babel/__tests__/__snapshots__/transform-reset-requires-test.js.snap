// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformPlugin inserts before each at top 1`] = `
"### Before

const React = require('react');
describe('test', () => {

});


### After
let React;
beforeEach(() => {
  React = require("react");
});
describe('test', () => {});"
`;

exports[`transformPlugin transforms act correctly 1`] = `
"### Before
const act = require('internal-test-utils').act;

### After
let act;
beforeEach(() => {
  act = require("internal-test-utils").act;
});"
`;

exports[`transformPlugin transforms all imports correctly 1`] = `
"### Before

const React = require('react');
const ReactDOM = require('react-dom');
const ReactDOMClient = require('react-dom/client');
const act = require('internal-test-utils').act;


### After
let act;
let ReactDOMClient;
let ReactDOM;
let React;
beforeEach(() => {
  React = require("react");
  ReactDOM = require("react-dom");
  ReactDOMClient = require("react-dom/client");
  act = require("internal-test-utils").act;
});"
`;

exports[`transformPlugin transforms babel output correctly 1`] = `
"### Before

var React = require('react');var _React = React,
startTransition = _React.startTransition,useDeferredValue = _React.useDeferredValue;
var ReactNoop = require('react-noop-renderer');var _require = require('internal-test-utils'),waitFor = _require.waitFor,waitForAll = _require.waitForAll,waitForPaint = _require.waitForPaint,waitForThrow = _require.waitForThrow,assertLog = _require.assertLog;
var act = require('internal-test-utils').act;


### After
let assertLog;
let waitForThrow;
let waitForPaint;
let waitForAll;
let waitFor;
let useDeferredValue;
let startTransition;
let act;

let _React;

let _require;

let ReactNoop;
let React;
beforeEach(() => {
  React = require("react");
  ReactNoop = require("react-noop-renderer");
  _require = require("internal-test-utils");
  _React = React;
  act = require("internal-test-utils").act;
  startTransition = _React.startTransition;
  useDeferredValue = _React.useDeferredValue;
  waitFor = _require.waitFor;
  waitForAll = _require.waitForAll;
  waitForPaint = _require.waitForPaint;
  waitForThrow = _require.waitForThrow;
  assertLog = _require.assertLog;
});"
`;

exports[`transformPlugin transforms destructured imports correctly 1`] = `
"### Before

const React = require('react');
const {startTransition, useDeferredValue} = React;
const {
  waitFor,
  waitForAll,
  waitForPaint,
  waitForThrow,
  assertLog,
  act,
} = require('internal-test-utils');


### After
let act;
let assertLog;
let waitForThrow;
let waitForPaint;
let waitForAll;
let waitFor;
let useDeferredValue;
let startTransition;
let React;
beforeEach(() => {
  React = require("react");
  startTransition = React.startTransition;
  useDeferredValue = React.useDeferredValue;
  waitFor = require("internal-test-utils").waitFor;
  waitForAll = require("internal-test-utils").waitForAll;
  waitForPaint = require("internal-test-utils").waitForPaint;
  waitForThrow = require("internal-test-utils").waitForThrow;
  assertLog = require("internal-test-utils").assertLog;
  act = require("internal-test-utils").act;
});"
`;

exports[`transformPlugin transforms require('react') correctly 1`] = `
"### Before
const React = require('react');

### After
let React;
beforeEach(() => {
  React = require("react");
});"
`;

exports[`transformPlugin transforms require('react-dom') correctly 1`] = `
"### Before
const ReactDOM = require('react-dom');

### After
let ReactDOM;
beforeEach(() => {
  ReactDOM = require("react-dom");
});"
`;

exports[`transformPlugin transforms require('react-dom/client') correctly 1`] = `
"### Before
const ReactDOMClient = require('react-dom/client');

### After
let ReactDOMClient;
beforeEach(() => {
  ReactDOMClient = require("react-dom/client");
});"
`;
